core/src/main/java/org/apache/calcite/rel/externalize/RelDotWriter.java

```diff
@@ -0,0 +1,210 @@
+
+  public RelDotWriter(
+      PrintWriter pw, SqlExplainLevel detailLevel,
+      boolean withIdPrefix) {
+    this(pw, detailLevel, withIdPrefix, new WriteOption());
+  }
+
+  public RelDotWriter(
+      PrintWriter pw, SqlExplainLevel detailLevel,
+      boolean withIdPrefix, WriteOption option) {
+    super(pw, detailLevel, withIdPrefix);
+    this.option = option;
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  protected void explain_(RelNode rel,
+                          List<Pair<String, Object>> values) {
+    // get inputs
+    List<RelNode> inputs = getInputs(rel);
+    outArcTable.put(rel, inputs);
+
+    // generate node label
+    StringBuilder relDesc = new StringBuilder();
+    populateRelNodeLabel(rel, values, relDesc);
+    nodeLabels.put(rel, option.formatNodeLabel(relDesc.toString()));
+
+    if (option.highlightNode(rel)) {
+      nodeStyles.put(rel, "bold");
+    }
+
+    explainInputs(inputs);
+  }
+
+  private List<RelNode> getInputs(RelNode parent) {
+    return parent.getInputs().stream().map(child -> {
+      if (child instanceof HepRelVertex) {
+        return ((HepRelVertex) child).getCurrentRel();
+      } else if (child instanceof RelSubset) {
+        RelSubset subset = (RelSubset) child;
+        return Util.first(subset.getBest(), subset.getOriginal());
+      } else {
+        return child;
+      }
+    }).collect(Collectors.toList());
+  }
+
+  private void explainInputs(List<RelNode> inputs) {
+    for (RelNode input : inputs) {
+      if (input == null || nodeLabels.containsKey(input)) {
+        continue;
+      }
+      input.explain(this);
+    }
+  }
+
+  public RelWriter done(RelNode node) {
+    int numOfVisitedNodes = nodeLabels.size();
+    super.done(node);
+    if (numOfVisitedNodes == 0) {
+      // When we enter this method call, no node
+      // has been visited. So the current node must be the root of the plan.
+      // Now we are exiting the method, all nodes in the plan
+      // have been visited, so it is time to dump the plan
+
+      pw.println("digraph {");
+
+      // print nodes with styles
+      for (RelNode rel : nodeStyles.keySet()) {
+        String style = String.join(",", nodeStyles.get(rel));
+        pw.println(nodeLabels.get(rel) + " [style=\"" + style + "\"]");
+      }
+
+      // ordinary arcs
+      for (Map.Entry<RelNode, List<RelNode>> entry : outArcTable.entrySet()) {
+        RelNode src = entry.getKey();
+        String srcDesc = nodeLabels.get(src);
+        for (int i = 0; i < entry.getValue().size(); i++) {
+          RelNode dst = entry.getValue().get(i);
+
+          // label is the ordinal of the arc
+          pw.println(srcDesc + " -> " + nodeLabels.get(dst) + " [label=\"" + i + "\"]");
+        }
+      }
+      pw.println("}");
+      pw.flush();
+    }
+    return this;
+  }
+
+  /**
+   * Options for displaying the rel node plan in dot format.
+   */
+  public static class WriteOption {
+
+    /**
```

Comment:
Did you try the `Config` as an `ImmutableBeans` ?

Reply:
Sounds reasonable. I have revised the PR so that each configuration item is immutable.

Reply:
I mean you can use a `ImmutableBeans` as configuration code instead of a specific class `WriteOption` here, an interface is enough, see how we do to `SqlValidatorImpl` .

Reply:
Revised accordingly. Thanks for your clarification.